tpm2/structures.go

```diff
@@ -464,13 +479,60 @@ func (ad AttestationData) Encode() ([]byte, error) {
 	if err != nil {
 		return nil, fmt.Errorf("encoding ExtraData, ClockInfo, FirmwareVersion: %v", err)
 	}
-	info, err := ad.AttestedCertifyInfo.encode()
-	if err != nil {
-		return nil, fmt.Errorf("encoding AttestedCertifyInfo: %v", err)
+
+	var info []byte
+	switch ad.Type {
+	case TagAttestCertify:
+		info, err = ad.AttestedCertifyInfo.encode()
+		if err != nil {
+			return nil, fmt.Errorf("encoding AttestedCertifyInfo: %v", err)
+		}
+	case TagAttestCreation:
+		info, err = ad.AttestedCreationInfo.encode()
+		if err != nil {
+			return nil, fmt.Errorf("encoding AttestedCreationInfo: %v", err)
+		}
 	}
+
 	return concat(head, signer, tail, info)
 }
 
+// CreationInfo contains Creation-specific data for TPMS_ATTEST.
+type CreationInfo struct {
+	Name         Name
+	OpaqueDigest []byte
```

Comment:
Should this be *HashValue?

Reply:
Negative. While most `TPM2B_Digest` structures contain a `TPMU_HA` structure, there is no requirement for them to, just the requirement that the structure cannot be longer than the longest digest size. (I spent 4 hours this morning finding this out the hard way 💀💀lol whelp)

Reply:
Can you add a comment here explaining that? Do you need access to this field when it's something else?

Maybe this could be something like

```
type CreationInfo struct {
    Name Name

    // Digest of the creation info. While common to many implementations, a digest isn't required, and may
    // be nil if the digest data isn't of the correct form.
    Digest *HashValue
    // RawDigest is the unparsed digest. This will always be present and can be used when implementations
    // don't use a TPMU_HA for the digest to parse alternative structures.
    RawDigest []byte
}
```

Reply:
Done.